fix: Improve spec compliance by improve handling of responses without content-length or transfer-encoding

evanderkoogh opened the pull request:
This fixes #1414 #1412 #1490

From [RFC 7230 3.3.3](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.3)
```
       Otherwise, this is a response message without a declared message
       body length, so the message body length is determined by the
       number of octets received prior to the server closing the
       connection.
```

So now if we get a SocketClosed, we check if we have seen a `content-length` or `transfer-encoding` header and if not, we complete the message parsing.

ronag reviewed (commented):
Needs a test. Also I need to think about this. Feels like there might be some edge cases here.Especially with pipelining.

evanderkoogh commented:
It does have a test: https://github.com/nodejs/undici/pull/1540/commits/ac2fbb96e612b6cbb999a956fbe03a66673d937f#diff-ecc2f3b14c967f02fd074ac2d0e0b876e683c84f2755d251bbaa74a813a82130

It is really weird that it says [line 620](https://app.codecov.io/gh/nodejs/undici/compare/1540/diff?src=pr&el=tree#D1L620) is not covered by test because [line 928](https://app.codecov.io/gh/nodejs/undici/compare/1540/diff?src=pr&el=tree#D1L968) can never be executed if `this.hasTransferEncoding` has not been set to true?

evanderkoogh commented:
Oh and the `Node CI / Test / test (18, ubuntu-latest, experimental) (pull_request) Failing after 2m` failing test is a timeout in some unrelated code.

Ahh.. I think I see where your concern is around the edge cases and pipelining.. I would need to check, but it could be indeed possible that when you are pipelining, it would error out the next request. I am going to investigate.

mcollina commented:
I think we should just error all the pipelined requests. Given they are GETs, they will be picked up whenever another socket is connected.

mcollina reviewed (changes requested):
Could you add a test with more than one `client.request()` executing in parallel with pipelining enabled and disabled? So we verify this works ok in all situations.

Guilledll commented:
Hi everyone, any news about merging this? It will be very usefull for me to have it, thanks for the great work.

mcollina commented:
This still needs one more test: https://github.com/nodejs/undici/pull/1540#pullrequestreview-1038547402

@evanderkoogh are you still interesting in completing this PR?

kyranet commented on lib/client.js:
```diff
@@ -615,6 +616,8 @@ class Parser {
       this.keepAlive += buf.toString()
     } else if (key.length === 14 && key.toString().toLowerCase() === 'content-length') {
       this.contentLength += buf.toString()
+    } else if (key.length === 16 && key.toString().toLowerCase() === 'transfer-encoding') {
```
Judging by the line 617 (since I'm not very familiar with Undici's code), shouldn't this be `17`? `transfer-encoding` has 3 more characters than the previous header, and `.length` does indeed evaluate to `17`.

ronag replied:
Sounds like a test is missing for this?

ronag commented on lib/client.js:
```diff
@@ -957,10 +960,16 @@ function onSocketEnd () {
+  }
+
+  parser.destroy()
+  parser = null
```
This doesn't properly null the parser. Please don't use a local var.

mcollina commented:
Would you mind sending a PR finishing this?

DamianGlowala commented:
Affects me (a Nuxt user) as well; receiving an error when proxying the request through a Nitro server:

![image](https://user-images.githubusercontent.com/48835293/212774736-e0c8d83e-9741-414c-b6ff-0868c2f48928.png)

garand commented:
😢 So is this not getting fix for now?

xgqfrms commented:
Any updates?

metcoder95 commented:
PR has been closed for a while, but this seems to be still something to be considered to add. If interested, feel free to open a new PR with a proposal and we can outline something 🙂

evanderkoogh commented:
The problem is that when I made that change, I fixed the actual issue, but then realised I introduced a bunch of edge-cases where there were still problems and I didn't have any bandwidth to figure out how to solve for all of those.

DamianGlowala commented:
I would have assumed a non-compliance with [RFC 7230 3.3.3](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.3) should gain a bit more attention from the Node.js team than simply 'feel free to open a new PR'...

Clearly this might be not a trivial problem to solve, which author of this PR already expressed in the above comment, hence any form of *support* from the team would be (very!) appreciated.

metcoder95 commented:
The team has limited bandwidth for addressing all the issues opened at the time, hence we asked the community if they are willing to contribute to fixing the problem meanwhile we provide all the support and guidance through reviews.

Since the PR has been closed for a while, it might be better to open a new one where we can continue the efforts if someone volunteers 🙂
